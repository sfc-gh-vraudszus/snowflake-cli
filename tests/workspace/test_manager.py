# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

import logging
import os
from pathlib import Path
from unittest import mock

import pytest
from snowflake.cli._plugins.workspace.manager import WorkspaceManager
from snowflake.cli.api.entities.common import EntityActions
from snowflake.cli.api.exceptions import InvalidProjectDefinitionVersionError
from snowflake.cli.api.project.definition_manager import DefinitionManager

from tests.nativeapp.patch_utils import mock_connection
from tests.testing_utils.files_and_dirs import create_named_file
from tests.workspace.utils import (
    APP_PACKAGE_ENTITY,
    MOCK_SNOWFLAKE_YML_FILE,
    MOCK_SNOWFLAKE_YML_V1_FILE,
)


@mock_connection()
def _get_ws_manager(mock_connection, pdf_content=MOCK_SNOWFLAKE_YML_FILE):
    current_working_directory = os.getcwd()
    create_named_file(
        file_name="snowflake.yml",
        dir_name=current_working_directory,
        contents=[pdf_content],
    )
    dm = DefinitionManager()
    return WorkspaceManager(
        project_definition=dm.project_definition,
        project_root=dm.project_root,
    )


def test_pdf_not_v2(temp_dir):
    with pytest.raises(InvalidProjectDefinitionVersionError):
        _get_ws_manager(pdf_content=MOCK_SNOWFLAKE_YML_V1_FILE)


# Test that the same entity instance is returned for the same id
def test_get_entity_is_cached(temp_dir):
    ws_manager = _get_ws_manager()
    pkg1 = ws_manager.get_entity("pkg")
    pkg2 = ws_manager.get_entity("pkg")
    app = ws_manager.get_entity("app")
    assert pkg1 is pkg2
    assert app is not pkg1


def test_get_entity_invalid_id(temp_dir):
    ws_manager = _get_ws_manager()
    with pytest.raises(ValueError, match="No such entity ID"):
        ws_manager.get_entity("non_existing_id")


def test_bundle(temp_dir):
    ws_manager = _get_ws_manager()
    with mock.patch(f"{APP_PACKAGE_ENTITY}.action_bundle") as app_pkg_bundle_mock:
        ws_manager.perform_action("pkg", EntityActions.BUNDLE)
        app_pkg_bundle_mock.assert_called_once()


def test_bundle_of_invalid_entity_type(temp_dir):
    ws_manager = _get_ws_manager()
    with pytest.raises(
        ValueError, match='This entity type does not support "action_bundle"'
    ):
        ws_manager.perform_action("app", EntityActions.BUNDLE)


# PJ-comment: what do you actually need to test here? just the version number top of file?
# PJ-comment: misleading to think everything in v2 is tested here
def test_migration_already_v2(
    runner,
    project_directory,
):
    with project_directory("migration_already_v2"):
        result = runner.invoke(["ws", "migrate"])

    assert result.exit_code == 0
    assert "Project definition is already at version 2." in result.output


def test_migrations_with_multiple_entities(
    runner, project_directory, os_agnostic_snapshot
):
    with project_directory("migration_multiple_entities"):
        result = runner.invoke(["ws", "migrate"])
    assert result.exit_code == 0
    assert Path("snowflake.yml").read_text() == os_agnostic_snapshot
    assert Path("snowflake_V1.yml").read_text() == os_agnostic_snapshot


# PJ-question: we are testing that the warning gets thrown here
@pytest.mark.parametrize(
    "project_directory_name", ["snowpark_templated_v1", "streamlit_templated_v1"]
)
def test_if_template_is_not_rendered_during_migration_with_option_checked(
    runner, project_directory, project_directory_name, os_agnostic_snapshot, caplog
):
    with project_directory(project_directory_name):
        with caplog.at_level(logging.WARNING):
            result = runner.invoke(["ws", "migrate", "--accept-templates"])

    assert result.exit_code == 0
    assert Path("snowflake.yml").read_text() == os_agnostic_snapshot
    assert Path("snowflake_V1.yml").read_text() == os_agnostic_snapshot
    assert (
        "Your V1 definition contains templates. We cannot guarantee the correctness of the migration."
        in caplog.text
    )


@pytest.mark.parametrize(
    "project_directory_name", ["snowpark_templated_v1", "streamlit_templated_v1"]
)
def test_if_template_raises_error_during_migrations(
    runner, project_directory, project_directory_name, os_agnostic_snapshot
):
    with project_directory(project_directory_name):
        result = runner.invoke(["ws", "migrate"])
        assert result.exit_code == 1
        assert "Project definition contains templates" in result.output


def test_migration_with_only_envs(project_directory, runner):
    with project_directory("sql_templating"):
        result = runner.invoke(["ws", "migrate"])

    assert result.exit_code == 0


@pytest.mark.parametrize(
    "duplicated_entity",
    [
        """
    - name: test
      handler: "test"
      signature: ""
      returns: string
      runtime: "3.10"
    """,
        """
streamlit:
  name: test
  stage: streamlit
  query_warehouse: test_warehouse
  main_file: "streamlit_app.py"
  title: "My Fancy Streamlit"
    """,
        """
    - name: test
      handler: "test"
      signature: ""
      returns: string
      handler: test
      runtime: "3.10"
    """,
    ],
)
def test_migrating_a_file_with_duplicated_keys_raises_an_error(
    runner, project_directory, os_agnostic_snapshot, duplicated_entity
):
    with project_directory("snowpark_procedures") as pd:
        definition_path = pd / "snowflake.yml"

        with open(definition_path, "a") as definition_file:
            definition_file.write(duplicated_entity)

        result = runner.invoke(["ws", "migrate"])
    assert result.exit_code == 1
    assert result.output == os_agnostic_snapshot
