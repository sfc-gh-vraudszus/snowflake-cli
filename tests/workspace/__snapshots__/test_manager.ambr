# serializer version: 1
# name: test_if_template_is_not_rendered_during_migration_with_option_checked[snowpark_templated_v1]
  '''
  definition_version: '2'
  entities:
    hello_function:
      artifacts:
      - dest: <! project_name | to_snowflake_identifier !>
        src: <% ctx.env.project_source %>
      external_access_integrations: []
      handler: functions.hello_function
      identifier:
        name: hello_function
      imports: []
      meta:
        use_mixins:
        - snowpark_shared
      returns: string
      secrets: {}
      signature:
      - name: name
        type: string
      stage: <! stage | to_snowflake_identifier !>
      type: function
    hello_procedure:
      artifacts:
      - dest: <! project_name | to_snowflake_identifier !>
        src: <% ctx.env.project_source %>
      execute_as_caller: false
      external_access_integrations: []
      handler: procedures.hello_procedure
      identifier:
        name: hello_procedure
      imports: []
      meta:
        use_mixins:
        - snowpark_shared
      returns: string
      secrets: {}
      signature:
      - name: name
        type: string
      stage: <! stage | to_snowflake_identifier !>
      type: procedure
    test_procedure:
      artifacts:
      - dest: <! project_name | to_snowflake_identifier !>
        src: <% ctx.env.project_source %>
      execute_as_caller: false
      external_access_integrations: []
      handler: procedures.test_procedure
      identifier:
        name: test_procedure
      imports: []
      meta:
        use_mixins:
        - snowpark_shared
      returns: string
      secrets: {}
      signature: ''
      stage: <! stage | to_snowflake_identifier !>
      type: procedure
  env:
    project_source: app/
  mixins:
    snowpark_shared:
      artifacts:
      - dest: <! project_name | to_snowflake_identifier !>
        src: <% ctx.env.project_source %>
      stage: <! stage | to_snowflake_identifier !>
  
  '''
# ---
# name: test_if_template_is_not_rendered_during_migration_with_option_checked[snowpark_templated_v1].1
  '''
  definition_version: "1.1"
  snowpark:
    project_name: "<! project_name | to_snowflake_identifier !>"
    stage_name: "<! stage | to_snowflake_identifier !>"
    src: <% ctx.env.project_source %>
    functions:
      - name: hello_function
        handler: "functions.hello_function"
        signature:
          - name: "name"
            type: "string"
        returns: string
    procedures:
      - name: hello_procedure
        handler: "procedures.hello_procedure"
        signature:
          - name: "name"
            type: "string"
        returns: string
      - name: test_procedure
        handler: "procedures.test_procedure"
        signature: ""
        returns: string
  env:
    project_source: "app/"
  
  '''
# ---
# name: test_if_template_is_not_rendered_during_migration_with_option_checked[streamlit_templated_v1]
  '''
  definition_version: '2'
  entities:
    streamlit_entity_1:
      artifacts:
      - streamlit_app.py
      - environment.yml
      - pages
      - common/hello.py
      identifier:
        name: <! name | to_snowflake_identifier !>
      main_file: streamlit_app.py
      pages_dir: pages
      query_warehouse: <! query_warehouse | to_snowflake_identifier !>
      stage: <! stage | to_snowflake_identifier !>
      title: <% ctx.env.streamlit_title %>
      type: streamlit
  env:
    streamlit_title: My Fancy Streamlit
  
  '''
# ---
# name: test_if_template_is_not_rendered_during_migration_with_option_checked[streamlit_templated_v1].1
  '''
  definition_version: "1.1"
  streamlit:
    name: <! name | to_snowflake_identifier !>
    stage: <! stage | to_snowflake_identifier !>
    query_warehouse: <! query_warehouse | to_snowflake_identifier !>
    main_file: streamlit_app.py
    env_file: environment.yml
    pages_dir: pages/
    title: <% ctx.env.streamlit_title %>
    additional_source_files:
      - common/hello.py
  env:
    streamlit_title: "My Fancy Streamlit"
  
  '''
# ---
# name: test_migrating_a_file_with_duplicated_keys_raises_an_error[\n    - name: test\n      handler: "test"\n      signature: ""\n      returns: string\n      handler: test\n      runtime: "3.10"\n    ]
  '''
  +- Error ----------------------------------------------------------------------+
  | While loading the project definition file, duplicate key was found: handler  |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_migrating_a_file_with_duplicated_keys_raises_an_error[\n    - name: test\n      handler: "test"\n      signature: ""\n      returns: string\n      runtime: "3.10"\n    ]
  '''
  +- Error ----------------------------------------------------------------------+
  | Entity with name test seems to be duplicated. Please rename it and try       |
  | again.                                                                       |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_migrating_a_file_with_duplicated_keys_raises_an_error[\nstreamlit:\n  name: test\n  stage: streamlit\n  query_warehouse: test_warehouse\n  main_file: "streamlit_app.py"\n  title: "My Fancy Streamlit"\n    ]
  '''
  +- Error ----------------------------------------------------------------------+
  | In your project, streamlit and snowpark entities share the same name. Please |
  | rename them and try again.                                                   |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_migrations_with_multiple_entities
  '''
  definition_version: '2'
  entities:
    app:
      from:
        target: pkg
      identifier: myapp_app
      meta:
        warehouse: app_wh
      type: application
    func1:
      artifacts:
      - dest: my_snowpark_project
        src: app
      external_access_integrations: []
      handler: app.func1_handler
      identifier:
        name: func1
      imports: []
      meta:
        use_mixins:
        - snowpark_shared
      returns: string
      runtime: '3.10'
      secrets: {}
      signature:
      - default: default value
        name: a
        type: string
      - name: b
        type: variant
      stage: dev_deployment
      type: function
    pkg:
      artifacts:
      - dest: ./
        src: app/*
      - dest: ./
        processors:
        - name: native app setup
        - name: templates
          properties:
            foo: bar
        src: to_process/*
      bundle_root: output/bundle/
      deploy_root: output/deploy/
      distribution: external
      generated_root: __generated/
      identifier: myapp_pkg
      manifest: app/manifest.yml
      meta:
        role: pkg_role
      scratch_stage: app_src.scratch
      stage: app_src.stage
      type: application package
    procedureName:
      artifacts:
      - dest: my_snowpark_project
        src: app
      execute_as_caller: false
      external_access_integrations: []
      handler: hello
      identifier:
        name: procedureName
      imports: []
      meta:
        use_mixins:
        - snowpark_shared
      returns: string
      secrets: {}
      signature:
      - name: name
        type: string
      stage: dev_deployment
      type: procedure
    test_streamlit:
      artifacts:
      - streamlit_app.py
      - environment.yml
      - pages
      identifier:
        name: test_streamlit
      main_file: streamlit_app.py
      pages_dir: None
      query_warehouse: test_warehouse
      stage: streamlit
      title: My Fancy Streamlit
      type: streamlit
  mixins:
    snowpark_shared:
      artifacts:
      - dest: my_snowpark_project
        src: app/
      stage: dev_deployment
  
  '''
# ---
# name: test_migrations_with_multiple_entities.1
  '''
  definition_version: 1
  streamlit:
    name: test_streamlit
    stage: streamlit
    query_warehouse: test_warehouse
    main_file: "streamlit_app.py"
    title: "My Fancy Streamlit"
  snowpark:
    project_name: "my_snowpark_project"
    stage_name: "dev_deployment"
    src: "app/"
    functions:
      - name: func1
        handler: "app.func1_handler"
        signature:
          - name: "a"
            type: "string"
            default: "default value"
          - name: "b"
            type: "variant"
        returns: string
        runtime: 3.10
    procedures:
      - name: procedureName
        handler: "hello"
        signature:
          - name: "name"
            type: "string"
        returns: string
  native_app:
    name: myapp
    source_stage: app_src.stage
    scratch_stage: app_src.scratch
    artifacts:
      - src: app/*
        dest: ./
      - src: to_process/*
        dest: ./
        processors:
          - native app setup
          - name: templates
            properties:
              foo: bar
    package:
      role: pkg_role
      distribution: external
    application:
      name: myapp_app
      warehouse: app_wh
      debug: true
  
  '''
# ---
